File: src/service/gospel.py
from __future__ import annotations

import json
import re
from copy import deepcopy
from datetime import datetime
from pathlib import Path
from zoneinfo import ZoneInfo

DEFAULT_TIMEZONE = "America/Edmonton"
ROOT = Path(__file__).resolve().parents[2]
READINGS_FILE = ROOT / "data" / "lectionary" / "readings.json"
DATE_RE = re.compile(r"^\d{4}-\d{2}-\d{2}$")


class InvalidDateError(ValueError):
    """Raised when date is not in YYYY-MM-DD format or is not a real date."""


class GospelNotFoundError(KeyError):
    """Raised when readings.json has no record for the requested date."""


def _resolve_date(date: str | None, timezone: str, today: str | None) -> str:
    if date is None:
        return today or datetime.now(ZoneInfo(timezone)).date().isoformat()
    if not DATE_RE.match(date):
        raise InvalidDateError("date must be YYYY-MM-DD")
    try:
        datetime.strptime(date, "%Y-%m-%d")
    except ValueError as exc:
        raise InvalidDateError("date must be a real YYYY-MM-DD date") from exc
    return date


def _load_readings(readings_file: Path = READINGS_FILE) -> dict:
    return json.loads(readings_file.read_text())


def get_gospel(
    date: str | None = None,
    *,
    timezone: str = DEFAULT_TIMEZONE,
    today: str | None = None,
    readings_file: Path = READINGS_FILE,
) -> dict:
    """Return citation-only Gospel metadata for a date.

    If date is omitted, today's date is calculated in the requested timezone.
    The returned payload intentionally contains no Bible text.
    """
    date_key = _resolve_date(date, timezone, today)
    readings = _load_readings(readings_file)
    entries = readings.get(date_key)
    if not entries:
        raise GospelNotFoundError(date_key)

    records = []
    for entry in entries:
        gospels = deepcopy(entry.get("readings", {}).get("gospel", []))
        records.append(
            {
                "feast": entry.get("feast"),
                "mass": entry.get("mass", "default"),
                "lectionary_number": entry.get("lectionary_number"),
                "gospels": gospels,
            }
        )

    return {
        "date": date_key,
        "timezone": timezone,
        "source": "catholic-daily-readings",
        "records": records,
    }

File: worker/src/gospel.ts
import readings from "../../data/lectionary/readings.json";

export const DEFAULT_TIMEZONE = "America/Edmonton";
const DATE_RE = /^\d{4}-\d{2}-\d{2}$/;

export class InvalidDateError extends Error {
  constructor(message = "date must be YYYY-MM-DD") {
    super(message);
    this.name = "InvalidDateError";
  }
}

type ReadingItem = {
  citation: string;
  sources: string[];
};

type DayEntry = {
  feast: string;
  mass?: string;
  lectionary_number: number | null;
  readings: {
    gospel?: ReadingItem[];
  };
};

export type GospelRecord = {
  feast: string;
  mass: string;
  lectionary_number: number | null;
  gospels: ReadingItem[];
};

export type GospelPayload = {
  date: string;
  timezone: string;
  source: "catholic-daily-readings";
  records: GospelRecord[];
};

type GetGospelOptions = {
  timezone?: string;
  today?: string;
};

function todayInTimeZone(timezone: string): string {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: timezone,
    year: "numeric",
    month: "2-digit",
    day: "2-digit",
  }).formatToParts(new Date());
  const values = Object.fromEntries(parts.map((part) => [part.type, part.value]));
  return `${values.year}-${values.month}-${values.day}`;
}

function resolveDate(date: string | undefined, timezone: string, today?: string): string {
  if (date === undefined || date === null || date === "") {
    return today ?? todayInTimeZone(timezone);
  }
  if (!DATE_RE.test(date)) {
    throw new InvalidDateError();
  }
  const parsed = new Date(`${date}T00:00:00Z`);
  if (Number.isNaN(parsed.getTime()) || parsed.toISOString().slice(0, 10) !== date) {
    throw new InvalidDateError("date must be a real YYYY-MM-DD date");
  }
  return date;
}

export function getGospel(date?: string, options: GetGospelOptions = {}): GospelPayload | null {
  const timezone = options.timezone ?? DEFAULT_TIMEZONE;
  const dateKey = resolveDate(date, timezone, options.today);
  const entries = (readings as Record<string, DayEntry[]>)[dateKey];
  if (!entries) {
    return null;
  }

  return {
    date: dateKey,
    timezone,
    source: "catholic-daily-readings",
    records: entries.map((entry) => ({
      feast: entry.feast,
      mass: entry.mass ?? "default",
      lectionary_number: entry.lectionary_number,
      gospels: entry.readings.gospel ?? [],
    })),
  };
}
